overlord/servicestate/quota_handlers.go

```diff
@@ -92,46 +93,145 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+		// restarting. There is a small chance that services
+		// will be restarted again but is preferable to the
+		// quota not applying to them.
+		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
+			return err
+		}
+
 	}
 
-	if err != nil {
+	if err := restartSnapServices(st, t, appsToRestartBySnap, perfTimings); err != nil {
 		return err
 	}
+	t.SetStatus(state.DoneStatus)
+	return nil
+}
 
-	// ensure service and slices on disk and their states are updated
-	opts := &ensureSnapServicesForGroupOptions{
-		allGrps: allGrps,
-	}
-	appsToRestartBySnap, err := ensureSnapServicesForGroup(st, t, grp, opts)
+var osutilBootID = osutil.BootID
+
+type quotaStateUpdated struct {
+	BootID              string              `json:"boot-id"`
+	AppsToRestartBySnap map[string][]string `json:"apps-to-restart,omitempty"`
+}
+
+func rememberQuotaStateUpdated(t *state.Task, appsToRestartBySnap map[*snap.Info][]*snap.AppInfo) error {
+	bootID, err := osutilBootID()
 	if err != nil {
 		return err
 	}
+	appNamesBySnapName := make(map[string][]string, len(appsToRestartBySnap))
+	for info, apps := range appsToRestartBySnap {
+		appNames := make([]string, len(apps))
+		for i, app := range apps {
+			appNames[i] = app.Name
+		}
+		appNamesBySnapName[info.InstanceName()] = appNames
+	}
+	t.Set("state-updated", quotaStateUpdated{
+		BootID:              bootID,
+		AppsToRestartBySnap: appNamesBySnapName,
+	})
+	return nil
+}
 
-	// after we have made all the persistent modifications to disk and state,
-	// set the task as done, what remains for this task handler is just to
-	// restart services which will happen regardless if we get rebooted after
-	// unlocking the state - if we got rebooted before unlocking the state, none
-	// of the changes we made to state would be persisted and we would run
-	// through everything above here again, but the second time around
-	// EnsureSnapServices would end up doing nothing since it is idempotent.
-	t.SetStatus(state.DoneStatus)
+func quotaStateAlreadyUpdated(t *state.Task) (ok bool, appsToRestartBySnap map[*snap.Info][]*snap.AppInfo, err error) {
+	var updated quotaStateUpdated
+	if err := t.Get("state-updated", &updated); err != nil {
+		if err == state.ErrNoState {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
 
-	return restartSnapServices(st, t, appsToRestartBySnap, perfTimings)
+	bootID, err := osutilBootID()
+	if err != nil {
+		return false, nil, err
+	}
+	if bootID != updated.BootID {
+		// rebooted => nothing to restart
+		return true, nil, nil
+	}
+
+	appsToRestartBySnap = make(map[*snap.Info][]*snap.AppInfo, len(updated.AppsToRestartBySnap))
+	st := t.State()
+	// best effort, ignore missing snaps and apps
+	for instanceName, appNames := range updated.AppsToRestartBySnap {
+		info, err := snapstate.CurrentInfo(st, instanceName)
+		if err != nil {
+			if _, ok := err.(*snap.NotInstalledError); ok {
+				t.Logf("snap %q missing over restart", instanceName)
```

Comment:
nitpick, this would be more clear as

```go
t.Logf("after snapd restart, snap %q went missing")
```

since we know in this code path that the only possible restart is that of snapd, not of the full system rebooting

Reply:
I changed this myself in https://github.com/snapcore/snapd/pull/10484